File: day03/day03.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func GetItemPriority(item rune) int {
	if item >= 'a' && item <= 'z' {
		return int(item) - int('a') + 1
	} else { // assume item >= 'A' and item <= 'Z'
		return int(item) - int('A') + 26 + 1
	}
}

func main() {
	dataFile, err := os.Open("input.txt")
	if err != nil {
		log.Fatal("failed to open data file")
	}

	var bags []string = make([]string, 0)

	fileScanner := bufio.NewScanner(dataFile)
	for fileScanner.Scan() {
		bags = append(bags, fileScanner.Text())
	}
	dataFile.Close()

	// Part 1
	totalPriorities := 0

	for _, bag := range bags {
		pivot := len(bag) / 2
		firstCompartment, secondComparment := bag[:pivot], bag[pivot:]
		var duplicatedItem rune

	out:
		for _, item := range firstCompartment {
			for _, item2 := range secondComparment {
				if item == item2 {
					duplicatedItem = item
					break out
				}
			}
		}
		totalPriorities += GetItemPriority(duplicatedItem)
	}

	// Part 2
	totalBadgePriorities := 0

	for bagIndex := 0; bagIndex < len(bags); bagIndex += 3 {
		var duplicatedItem rune

	out_3:
		for _, item := range bags[bagIndex+0] {
			for _, item2 := range bags[bagIndex+1] {
				if item2 != item {
					continue
				}
				for _, item3 := range bags[bagIndex+2] {
					if item3 == item {
						duplicatedItem = item
						break out_3
					}
				}
			}
		}
		totalBadgePriorities += GetItemPriority(duplicatedItem)
	}

	fmt.Printf(("Part 1: %d\n"), totalPriorities)
	fmt.Printf(("Part 2: %d\n"), totalBadgePriorities)
}

File: aoc2022/day03/day03.py
def get_priority(item):
    item = ord(item[0])
    if item >= ord('a') and item <= ord('z'):
        priority =  item - ord('a') + 1
        return priority
    elif item >= ord('A') and item <= ord('Z'):
        return  item - ord('A') + 1 + 26

    raise Exception('invalid item (not in [a-zA-Z]')
    return -1

bags = []
# with open('sample.txt','r') as data_file:
with open('input.txt','r') as data_file:
    bags = [ line.strip() for line in data_file.readlines() ]

total_priorities = 0

# Part 1
for bag in bags:
    pivot = len(bag)//2
    first_compartment, second_compartment = bag[:pivot], bag[pivot:]
    duplicated_item = [ item for item in first_compartment if item in second_compartment ]
    total_priorities += get_priority(duplicated_item)

# Part 2
total_badge_priorities = 0

for i in range(0, len(bags), 3):
    bp1 = bags[i]
    bp2 = bags[i+1]
    bp3 = bags[i+2]
    badge = [item2 for item2 in [ item for item in bp1 if item in bp2 ] if item2 in bp3][0]
    total_badge_priorities += get_priority(badge)


print(f'Part 1: {total_priorities}') # 7903

print(f'Part 2: {total_badge_priorities}') # 2548


